File: Backend/BackendTienda/BackendTienda/Feautures/Employees/EmployeeService.cs
using BackendTienda.Data;
using BackendTienda.Shared.Models;
using BackendTienda.Shared.Database;
using Microsoft.EntityFrameworkCore;

namespace BackendTienda.Feautures.Employees
{
    public class EmployeeService(AppDbContext _db)
    {
        public async Task<IEnumerable<GetEmployee>> GetAll()
        {
            var employees = await _db.employee.Include(p => p.Position).ToListAsync();

            return employees.Select(e => new GetEmployee(
                 employeeId: e.EmployeeId,
                  fullName: e.FullName,
                     email: e.Email,
                     birthDate: e.BirthDate,
                     positionId: e.PositionId,
                     position: e.Position.Name
                 )).ToList();
        }

        public async Task<GetEmployee> GetById(int Id)
        {
            var e = await _db.employee.Include(p => p.Position)
                .FirstAsync(e => e.EmployeeId == Id);

            return new GetEmployee(
                 employeeId: e.EmployeeId,
                  fullName: e.FullName,
                     email: e.Email,
                     birthDate: e.BirthDate,
                     positionId: e.PositionId,
                     position: e.Position.Name
                 );
        }

        public async Task Add(CreateEmployee request)
        {

            var employee = new Employee()
            {
                fullName = request.fullName,
                email = request.email,
                birthDate = request.birthDate,
                positionId = request.positionId
            };

            await _db.employee.AddAsync(employee);
            await _db.SaveChangesAsync();

        }

        public async Task Update(UpdateEmployee request)
        {
            var employeeFound = await _db.employee.FirstAsync(e => e.employeeId == request.employeeId);

            employeeFound.FullName = request.fullName;
            employeeFound.Email = request.email;
            employeeFound.BirthDate = request.birthDate;
            employeeFound.PositionId = request.positionId;

            await _db.SaveChangesAsync();

        }

        public async Task Delete(int id)
        {
            var employeeFound = await _db.employee.FirstAsync(e => e.employeeId == id);

            _db.employee.Remove(employeeFound);
            await _db.SaveChangesAsync();

        }

    }
}

File: Backend/BackendTiendaMigrado/backendtienda/src/main/java/com/backend/backendtienda/service/EmployeeService.java
package com.backend.backendtienda.service;

import com.backend.backendtienda.dto.EmployeeDTOs.CreateEmployee;
import com.backend.backendtienda.dto.EmployeeDTOs.GetEmployee;
import com.backend.backendtienda.dto.EmployeeDTOs.UpdateEmployee;
import com.backend.backendtienda.entity.Employee;
import com.backend.backendtienda.entity.Position;
import com.backend.backendtienda.repository.EmployeeRepository;
import com.backend.backendtienda.repository.PositionRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.List;

@Service
public class EmployeeService {

    private final EmployeeRepository employeeRepository;
    private final PositionRepository positionRepository;
    private final PasswordEncoder passwordEncoder;   

    public EmployeeService(EmployeeRepository employeeRepository,
                           PositionRepository positionRepository,
                           PasswordEncoder passwordEncoder) {   
        this.employeeRepository = employeeRepository;
        this.positionRepository = positionRepository;
        this.passwordEncoder = passwordEncoder;   
    }

    @Transactional(readOnly = true)
    public List<GetEmployee> getAll() {
        return employeeRepository.findAll().stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public GetEmployee getById(Integer id) {
        return toResponse(findEmployee(id));
    }

    @Transactional
    public void add(CreateEmployee req) {
        Employee employee = new Employee();
        employee.setFullName(req.fullName());
        employee.setEmail(req.email());
        employee.setBirthDate(req.birthDate());
        employee.setPosition(findPosition(req.positionId()));
        employee.setPassword(passwordEncoder.encode(req.password()));   // ← NUEVO

        employeeRepository.save(employee);
    }

    @Transactional
    public void update(UpdateEmployee req) {
    Employee employee = findEmployee(req.employeeId());
    employee.setFullName(req.fullName());
    employee.setEmail(req.email());
    employee.setBirthDate(req.birthDate());
    employee.setPosition(findPosition(req.positionId()));

    if (req.password() != null && !req.password().isBlank()) {  
        employee.setPassword(passwordEncoder.encode(req.password()));
    }
    
}

    @Transactional
    public void delete(Integer id) {
        employeeRepository.delete(findEmployee(id));
    }

    private Employee findEmployee(Integer id) {
        return employeeRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "El empleado no existe"));
    }

    private Position findPosition(Integer id) {
        if (id == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "El puesto es obligatorio");
        }
        return positionRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.BAD_REQUEST, "El puesto no existe"));
    }

    private GetEmployee toResponse(Employee e) {
        return new GetEmployee(
                e.getEmployeeId(),
                e.getFullName(),
                e.getEmail(),
                e.getBirthDate(),
                e.getPosition().getPositionId(),
                e.getPosition().getName());
    }
}